homeassistant/components/fritz/common.py

```diff
@@ -280,41 +305,75 @@ def scan_devices(self, now: datetime | None = None) -> None:
             consider_home = _default_consider_home
 
         new_device = False
-        for known_host in self._update_hosts_info():
-            if not known_host.get("mac"):
+        hosts = {}
+        for host in self._update_hosts_info():
+            if not host.get("mac"):
+                continue
+
+            hosts[host["mac"]] = Device(
+                name=host["name"],
+                connected=host["status"],
+                connected_to="",
+                connection_type="",
+                ip_address=host["ip"],
+                ssid=None,
+                wan_access=False,
+            )
+
+        mesh_intf = {}
+        # first get all meshed devices
+        for node in topology["nodes"]:
+            if not node["is_meshed"]:
                 continue
 
-            dev_mac = known_host["mac"]
-            dev_name = known_host["name"]
-            dev_ip = known_host["ip"]
-            dev_home = known_host["status"]
-            dev_wan_access = True
-            if dev_ip:
-                wan_access = self.connection.call_action(
-                    "X_AVM-DE_HostFilter:1",
-                    "GetWANAccessByIP",
-                    NewIPv4Address=dev_ip,
+            for interf in node["node_interfaces"]:
+                int_mac = interf["mac_address"]
+                mesh_intf[interf["uid"]] = Interface(
+                    device=node["device_name"],
+                    mac=int_mac,
+                    ssid=interf.get("ssid", ""),
+                    type=interf["type"],
                 )
-                if wan_access:
-                    dev_wan_access = not wan_access.get("NewDisallow")
+                if format_mac(int_mac) == format_mac(self.mac):
+                    self.mesh_role = node["mesh_role"]
```

Comment:
```suggestion
                    self.mesh_role = MeshRoles(node["mesh_role"])
```
when we already have a suitable enum, than we should use it

Reply:
;-)